source/quic_server_ssl_poll_http.c

```diff
@@ -0,0 +1,3286 @@
+        (head)->tqh_last = &(head)->tqh_first; \
+        (head)->tqh_last = &(head)->tqh_first; \
+        (head)->tqh_count = 0; \
+    } while (0)
+
+#define QPOLL_TAILQ_INSERT_HEAD(head, elm, field) \
+    do { \
+        if (((elm)->field.tqe_next = (head)->tqh_first) != NULL) \
+            (head)->tqh_first->field.tqe_prev = &(elm)->field.tqe_next; \
+        else \
+            (head)->tqh_last = &(elm)->field.tqe_next; \
+        (head)->tqh_first = (elm); \
+        (elm)->field.tqe_prev = &(head)->tqh_first; \
+        (head)->tqh_count++; \
+    } while (0)
+
+#define QPOLL_TAILQ_INSERT_TAIL(head, elm, field) \
+    do { \
+        (elm)->field.tqe_next = NULL; \
+        (elm)->field.tqe_prev = (head)->tqh_last; \
+        *(head)->tqh_last = (elm); \
+        (head)->tqh_last = &(elm)->field.tqe_next; \
+        (head)->tqh_count++; \
+    } while (0)
+
+#define QPOLL_TAILQ_REMOVE(head, elm, field) \
+    do { \
+        if (((elm)->field.tqe_next) != NULL) \
+            (elm)->field.tqe_next->field.tqe_prev = (elm)->field.tqe_prev; \
+        else \
+            (head)->tqh_last = (elm)->field.tqe_prev; \
+        *(elm)->field.tqe_prev = (elm)->field.tqe_next; \
+        (head)->tqh_count--; \
+    } while (0)
+
+/*
+ * every poll_event structure has members enumerated here in poll_event_base
+ * The poll events are kept in list which is managed by poll_manager instance.
+ * However SSL_poll(9ossl) expects an array of SSL_POLL_ITEM structures. Therefore
+ * the poll manager needs to construct array of poll_event structures from list.
+ * We get two copies of poll_event structures:
+ *    - one copy is held in list (original)
+ *    - one copy is held in array (copy for SSL_poll())
+ * We use pe_self member to easily reach from SSL_poll() copy instance back to
+ * original managed by poll manager and used by our application.
+ * There are four callbacks:
+ *    - pe_cb_in() - triggered when inbound data/stream/connection is coming
+ *    - pe_cb_out() - triggered when outbound data/stream/connection is coming
+ *    - pe_cb_error() - triggered when polled object enters an error state
+ *    - pe_cb_destroy() - this a destructor, application destroy pe_appdata
+ * The remaining members are rather self explanatory.
+ */
+#define poll_event_base \
+    SSL_POLL_ITEM pe_poll_item; \
+    QPOLL_TAILQ_ENTRY(poll_event) pe_tqe; \
+    uint64_t pe_want_events; \
+    uint64_t pe_want_mask; \
+    struct poll_manager *pe_my_pm; \
+    unsigned char pe_type; \
+    struct poll_event *pe_self; \
+    int(*pe_cb_in)(struct poll_event *); \
+    int(*pe_cb_out)(struct poll_event *); \
+    int(*pe_cb_error)(struct poll_event *); \
+    void(*pe_cb_ondestroy)(struct poll_event *)
+
+struct poll_event {
+    poll_event_base;
+};
+
+struct poll_event_listener {
+    poll_event_base;
+};
+
+/*
+ * The poll_event is associated with SSL object which is typically QUIC stream.
+ * There are two types of streams in QUIC:
```

Comment:
Kind of strange that you say there are two types of streams and then list four things in your bullet list (two of which are streams).